examples/knn_classification.py

```diff
@@ -0,0 +1,438 @@
+                        help='The batch size. If the input is variable sized, then this argument becomes a maximum.'
+    )
+    parser.add_argument('-w', '--workers', default=8, type=int,
+                        help='The number of data loader workers to use per GPU'
+    )
+    parser.add_argument('-k', default=20, type=int,
+                        help="How many neighbors to use for classification."
+    )
+
+    args, _ = parser.parse_known_args()
+
+    def rank_print(*args, **kwargs):
+        if rank == 0:
+            print(*args, **kwargs)
+
+    rank_print('Loading model...')
+    if args.use_hf:
+        from transformers import AutoModel
+        model = AutoModel.from_pretrained(f"nvidia/{args.model_version}", trust_remote_code=True)
+    else:
+        model = torch.hub.load('NVlabs/RADIO', 'radio_model', version=args.model_version, progress=True)
+
+    model.to(device=device).eval()
+    rank_print('Done')
+
+    transform = [
+        ResizeTransform(args.resolution, args.resize_multiple),
+    ]
+    if len(args.resolution) == 2:
+        transform.append(transforms.CenterCrop(args.resolution))
+    transform.extend([
+        transforms.ToImage(),
+        transforms.ToDtype(torch.float32, scale=True),
+    ])
+    transform = transforms.Compose(transform)
+
+    rank_print('Loading dataset...')
+    ds_train_builder = load_dataset_builder(args.dataset, trust_remote_code=True)
+    num_classes = ds_train_builder.info.features['label'].num_classes
+    num_train_examples = ds_train_builder.info.splits[args.train_split].num_examples
+
+    def _prepare(builder):
+        for i in range(min(2, world_size)):
+            if i == rank or (i > 0 and rank > 0):
+                builder.download_and_prepare()
+            if world_size > 1:
+                dist.barrier()
+    _prepare(ds_train_builder)
+
+    if args.eval_dataset:
+        ds_eval_builder = load_dataset_builder(args.eval_dataset, trust_remote_code=True)
+        _prepare(ds_eval_builder)
+    else:
+        ds_eval_builder = ds_train_builder
+
+    num_eval_examples = ds_eval_builder.info.splits[args.eval_split].num_examples
+    assert num_classes == ds_eval_builder.info.features['label'].num_classes, "The number of classes must match between train and eval!"
+
+    num_train_steps = _round_up(num_train_examples, args.batch_size * world_size)
+    num_eval_steps = _round_up(num_eval_examples, args.batch_size * world_size)
+
+    def _get_dataset(builder, split: str):
+        dataset = builder.as_dataset(split=split)
+        dataset = dataset.to_iterable_dataset(num_shards=world_size * max(1, args.workers))
+        dataset = split_dataset_by_node(dataset, rank=rank, world_size=world_size)
+        dataset = dataset.map(lambda ex: dict(image=transform(ex['image']), label=torch.as_tensor(ex['label'], dtype=torch.int64)))
+
+        loader = DataLoader(dataset, batch_size=args.batch_size, shuffle=False,
```

Comment:
Is it legal to use batch_size>1 when we don't do center crops?

Reply:
The collater groups the images into size bins. So "batch_size" becomes more fluid in the sense that we load that many examples, group them based on size, and then process those parts.